File: python_rtp/ServerWorker.py
# ServerWorker.py

from random import randint
import sys, traceback, threading, socket, time

from RtpPacket import RtpPacket

MAX_PAYLOAD = 1400
SOF = b'\xff\xd8'   # JPEG Start
EOF = b'\xff\xd9'   # JPEG End

class ServerWorker:
    SETUP = 'SETUP'
    PLAY = 'PLAY'
    PAUSE = 'PAUSE'
    TEARDOWN = 'TEARDOWN'

    INIT = 0
    READY = 1
    PLAYING = 2
    state = INIT

    OK_200 = 0
    FILE_NOT_FOUND_404 = 1
    CON_ERR_500 = 2

    clientInfo = {}

    def __init__(self, clientInfo):
        self.clientInfo = clientInfo

    def run(self):
        threading.Thread(target=self.recvRtspRequest).start()

    def recvRtspRequest(self):
        connSocket = self.clientInfo['rtspSocket'][0]
        while True:
            data = connSocket.recv(256)
            if data:
                print("RTSP Data:\n" + data.decode("utf-8"))
                self.processRtspRequest(data.decode("utf-8"))

    def processRtspRequest(self, data):
        request = data.split('\n')
        line1 = request[0].split(' ')
        requestType = line1[0]
        filename = line1[1]
        seq = request[1].split(' ')

        # SETUP 
        if requestType == self.SETUP:
            if self.state == self.INIT:
                print("processing SETUP\n")
                try:
                    self.loadMJPEG(filename)
                    self.clientInfo['pktSeq'] = 0
                    self.clientInfo['frameIndex'] = 0
                    self.state = self.READY
                except IOError:
                    self.replyRtsp(self.FILE_NOT_FOUND_404, seq[1])
                    return

                self.clientInfo['session'] = randint(100000, 999999)
                self.replyRtsp(self.OK_200, seq[1])

                try:
                    self.clientInfo['rtpPort'] = int(request[2].split('=')[1].strip())
                except:
                    self.clientInfo['rtpPort'] = 25000

        # PLAY 
        elif requestType == self.PLAY:
            if self.state == self.READY:
                print("processing PLAY\n")
                self.state = self.PLAYING

                self.clientInfo["rtpSocket"] = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                self.replyRtsp(self.OK_200, seq[1])

                self.clientInfo['event'] = threading.Event()
                self.clientInfo['worker'] = threading.Thread(target=self.sendRtp)
                self.clientInfo['worker'].start()

        # PAUSE 
        elif requestType == self.PAUSE:
            if self.state == self.PLAYING:
                print("processing PAUSE\n")
                self.state = self.READY
                self.clientInfo['event'].set()
                self.replyRtsp(self.OK_200, seq[1])

        # TEARDOWN 
        elif requestType == self.TEARDOWN:
            print("processing TEARDOWN\n")
            if 'event' in self.clientInfo:
                self.clientInfo['event'].set()
            self.replyRtsp(self.OK_200, seq[1])

            if 'rtpSocket' in self.clientInfo:
                try: self.clientInfo['rtpSocket'].close()
                except: pass

    # MJPEG READER – RAW FFD8 / FFD9 PARSING
    def loadMJPEG(self, filename):
        """Load MJPEG file & split all frames by JPEG markers."""
        with open(filename, "rb") as f:
            data = f.read()

        frames = []
        i = 0
        while True:
            start = data.find(SOF, i)
            if start < 0:
                break
            end = data.find(EOF, start)
            if end < 0:
                break
            end += 2  # include FFD9
            frames.append(data[start:end])
            i = end

        if len(frames) == 0:
            raise IOError("No JPEG frames found!")

        print(f"Loaded {len(frames)} frames from MJPEG.")
        self.clientInfo['frames'] = frames

    # RTP SENDER
    def sendRtp(self):
        """Send RTP packets repeatedly."""
        while True:
            self.clientInfo['event'].wait(0.04)  # 25 FPS

            if self.clientInfo['event'].isSet():
                break

            frames = self.clientInfo['frames']
            idx = self.clientInfo['frameIndex']

            frame = frames[idx]
            frameTimestamp = idx  # Simple timestamp = frame index

            # Move to next frame (loop)
            self.clientInfo['frameIndex'] = (idx + 1) % len(frames)

            # Fragment and send
            self.sendFrame(frame, frameTimestamp)

    def sendFrame(self, frame_bytes, timestamp):
        """Fragment frame into RTP packets."""
        total = len(frame_bytes)
        num = (total + MAX_PAYLOAD - 1) // MAX_PAYLOAD

        address = self.clientInfo['rtspSocket'][1][0]
        port = int(self.clientInfo['rtpPort'])

        for i in range(num):
            start = i * MAX_PAYLOAD
            end = start + MAX_PAYLOAD
            chunk = frame_bytes[start:end]

            marker = 1 if i == num - 1 else 0

            # Increase packet seq num
            self.clientInfo['pktSeq'] = (self.clientInfo['pktSeq'] + 1) & 0xFFFF

            packet = self.makeRtp(chunk, timestamp, self.clientInfo['pktSeq'], marker)
            self.clientInfo['rtpSocket'].sendto(packet, (address, port))

    def makeRtp(self, payload_chunk, timestamp, seqnum, marker):
        version = 2
        padding = 0
        extension = 0
        cc = 0
        pt = 26     # MJPEG
        ssrc = 12345

        pkt = RtpPacket()
        pkt.encode(
            version, padding, extension, cc,
            seqnum,
            timestamp,   # correct timestamp
            marker,
            pt,
            ssrc,
            payload_chunk
        )
        return pkt.getPacket()

    def replyRtsp(self, code, seq):
        if code == self.OK_200:
            reply = 'RTSP/1.0 200 OK\nCSeq: ' + str(seq) + '\nSession: ' + str(self.clientInfo['session'])
            connSocket = self.clientInfo['rtspSocket'][0]
            connSocket.send(reply.encode())
        elif code == self.FILE_NOT_FOUND_404:
            print("404 NOT FOUND")
        elif code == self.CON_ERR_500:
            print("500 CONNECTION ERROR")

File: java/ServerWorker.java
package java;

// ServerWorker.java
import java.net.*;
import java.io.*;
import java.util.Random;

public class ServerWorker extends Thread {

    // RTSP states
    static final int INIT = 0;
    static final int READY = 1;
    static final int PLAYING = 2;

    // RTSP reply codes
    static final int OK_200 = 0;
    static final int FILE_NOT_FOUND_404 = 1;
    static final int CON_ERR_500 = 2;

    Socket rtspSocket;
    BufferedReader rtspReader;
    BufferedWriter rtspWriter;

    InetAddress clientIP;
    int clientRTPPort = 25000;

    int state;
    int RTSPid;
    int RtspSeqNb = 0;

    VideoStream video;
    DatagramSocket rtpSocket = null;
    int RTPseq = 0;
    int framePeriod = 40; // ms (~25fps)

    volatile boolean teardown = false;
    volatile boolean sending = false;

    public ServerWorker(Socket client) throws IOException {
        this.rtspSocket = client;
        rtspReader = new BufferedReader(new InputStreamReader(rtspSocket.getInputStream()));
        rtspWriter = new BufferedWriter(new OutputStreamWriter(rtspSocket.getOutputStream()));
        clientIP = client.getInetAddress();
        state = INIT;
        RTSPid = new Random().nextInt(900000) + 100000;
    }

    public void run() {
        try {
            while (!teardown) {
                String request = recvRtspRequest();
                if (request == null)
                    break;
                processRtspRequest(request);
            }
        } catch (Exception e) {
            System.out.println("ServerWorker error: " + e.getMessage());
        } finally {
            cleanup();
        }
    }

    // Read RTSP request (read first line + up to several headers)
    private String recvRtspRequest() throws IOException {
        StringBuilder sb = new StringBuilder();
        String line = rtspReader.readLine();
        if (line == null)
            return null;
        sb.append(line).append("\n");
        // read next header lines (CSeq, Transport/Session...)
        for (int i = 0; i < 10; i++) {
            rtspSocket.setSoTimeout(50);
            String l;
            try {
                l = rtspReader.readLine();
            } catch (SocketTimeoutException ste) {
                l = null;
            }
            if (l == null || l.length() == 0)
                break;
            sb.append(l).append("\n");
        }
        return sb.toString();
    }

    private void processRtspRequest(String data) throws Exception {
        System.out.println("Received RTSP:\n" + data);
        String[] lines = data.split("\n");
        String[] requestLine = lines[0].split(" ");
        String requestType = requestLine[0];
        String fileName = requestLine.length > 1 ? requestLine[1] : "";
        String cseq = (lines.length > 1) ? lines[1].split(" ")[1] : "0";
        RtspSeqNb = Integer.parseInt(cseq);

        if (requestType.equals("SETUP") && state == INIT) {
            try {
                video = new VideoStream(fileName);
            } catch (IOException ioe) {
                replyRtsp(FILE_NOT_FOUND_404, cseq);
                return;
            }
            // parse Transport header for client port
            for (int i = 2; i < lines.length; i++) {
                if (lines[i].toLowerCase().contains("transport")) {
                    String transport = lines[i];
                    int idx = transport.indexOf("client_port=");
                    if (idx >= 0) {
                        String portPart = transport.substring(idx + "client_port=".length()).trim();
                        if (portPart.contains("-"))
                            portPart = portPart.split("-")[0];
                        try {
                            clientRTPPort = Integer.parseInt(portPart);
                        } catch (NumberFormatException e) {
                            clientRTPPort = 25000;
                        }
                    }
                }
            }

            // create DatagramSocket for sending RTP
            rtpSocket = new DatagramSocket();
            state = READY;
            replyRtsp(OK_200, cseq);

        } else if (requestType.equals("PLAY") && state == READY) {
            replyRtsp(OK_200, cseq);
            state = PLAYING;
            sending = true;
            Thread sender = new Thread(() -> {
                try {
                    sendRtp();
                } catch (Exception e) {
                    System.out.println("sendRtp error: " + e.getMessage());
                }
            });
            sender.start();

        } else if (requestType.equals("PAUSE") && state == PLAYING) {
            replyRtsp(OK_200, cseq);
            state = READY;
            sending = false;

        } else if (requestType.equals("TEARDOWN")) {
            replyRtsp(OK_200, cseq);
            teardown = true;
            sending = false;
        } else {
            replyRtsp(CON_ERR_500, cseq);
        }
    }

    private void sendRtp() throws Exception {
        byte[] frameBuf = new byte[2000000]; // large buffer for HD
        while (sending && !teardown) {
            int frameLen = video.getnextframe(frameBuf);
            if (frameLen <= 0) {
                sending = false;
                break;
            }
            RTPseq++;
            int timestamp = RTPseq; // simple timestamp
            RTPpacket rtpPacket = new RTPpacket(26, RTPseq, timestamp, frameBuf, frameLen);

            int packetLen = rtpPacket.getlength();
            byte[] packet = new byte[packetLen];
            rtpPacket.getpacket(packet);

            DatagramPacket dp = new DatagramPacket(packet, packet.length, clientIP, clientRTPPort);
            try {
                rtpSocket.send(dp);
            } catch (IOException ioe) {
                System.out.println("Connection Error while sending RTP: " + ioe.getMessage());
                sending = false;
                break;
            }

            try {
                Thread.sleep(framePeriod);
            } catch (InterruptedException ie) {
                break;
            }
        }
    }

    private void replyRtsp(int code, String cseq) {
        String reply;
        if (code == OK_200) {
            reply = "RTSP/1.0 200 OK\nCSeq: " + cseq + "\nSession: " + RTSPid + "\n";
        } else if (code == FILE_NOT_FOUND_404) {
            reply = "RTSP/1.0 404 NOT FOUND\nCSeq: " + cseq + "\nSession: " + RTSPid + "\n";
        } else {
            reply = "RTSP/1.0 500 ERROR\nCSeq: " + cseq + "\nSession: " + RTSPid + "\n";
        }
        try {
            rtspWriter.write(reply);
            rtspWriter.flush();
            System.out.println("Sent RTSP reply:\n" + reply);
        } catch (IOException e) {
            System.out.println("Failed to send RTSP reply: " + e.getMessage());
        }
    }

    private void cleanup() {
        try {
            if (video != null)
                video.close();
        } catch (Exception e) {
        }
        try {
            if (rtpSocket != null)
                rtpSocket.close();
        } catch (Exception e) {
        }
        try {
            if (rtspSocket != null)
                rtspSocket.close();
        } catch (Exception e) {
        }
    }
}